lab/skel_graph/astar/main.cpp

```diff
@@ -0,0 +1,228 @@
+
+
+        std::cin >> source.first >> source.second;
+        std::cin >> destination.first >> destination.second;
+    }
+
+
+    void print_output() {
+        for (auto &pos : build_path(destination)) {
+            std::cout<< "Trec prin punctul" << pos.first << " " << pos.second;
+            std::cout << "\n";
+        }
+    }
+
+    std::vector<std::pair <int, int>>  expand(std::pair <int, int> node) {
+        std::vector<std::pair <int, int>> directions =
+        {{1, 1}, {1, 0}, {1, -1}, {0, 1}, {0, -1}, {-1, 1}, {-1, 0}, {-1, -1},};
+        std::vector<std::pair <int, int>> succ;
+        for (const auto &[dx, dy] : directions) {
+            int x = node.first + dx;
+            int y = node.second + dy;
+            if (x >= 1 && x <= n && y >= 1 && y <= n && adj_matrix[x][y] == 1) {
+                succ.push_back({x, y});
+            }
+        }
+        return succ;
+    }
+
+    bool is_goal(const Cell &node) {
+        return (node == destination);
+    }
+
+    double euclidian_distance_heuristic(Cell node) {
+        int abs_x = abs(node.first - destination.first);
+        int abs_y = abs(node.second - destination.second);
+        return sqrt(abs_x * abs_x + abs_y * abs_y);
+    }
+
+    std::vector<std::pair <int, int>> build_path(Cell node) {
+        auto current = node;
+        std::vector<Cell> path;
+        while (!(current.first == -1 && current.second == -1)) {
+            path.push_back(current);
+            current = parent[current.first][current.second];
+        }
+        std::reverse(path.begin(), path.end());
+        return path;
+    }
+
+    bool astar(const std::pair<int, int> &start_state) {
+         // initializari
+        open.clear();  // multimea nodurilor in curs de explorare
+        closed.clear();  // multimea nodurilor expandate
+
+        // functie lambda pentru a compara doua perechi
+        auto comparator = [](auto x1, auto x2) {
+        return x1.first > x2.first;
+        };
+
+        //  priority queue cu nodurile in curs de explorare
+        std::priority_queue<std::pair<double, Cell>,
+        std::vector<std::pair <double, Cell>>,
+        decltype(comparator)> openlist(comparator);
+
+        // pentru s_i avem distanta 0 si parinte NULL
+        std::get<1>(fgh[start_state.first][start_state.second]) = 0.0;
+        std::get<2>(fgh[start_state.first][start_state.second]) =
+        euclidian_distance_heuristic(start_state);
+        std::get<0>(fgh[start_state.first][start_state.second]) =
+        std::get<1>(fgh[start_state.first][start_state.second])
+        + std::get<2>(fgh[start_state.first][start_state.second]);
+
+        // adaugare nod atat in lista de open, cat si in coada de prioritati.
+        openlist.push
+        (std::make_pair(std::get<0>
+        (fgh[start_state.first][start_state.second]), start_state));
+        open.insert(start_state);
+
+        // parcurgerea continua cat timp avem noduri in open
+        while (!openlist.empty()) {
```

Comment:
We don't need an `open` list. We need a C++ solution for storing only the `open`.

Reply:
@neatudarius I was thinking about using std::set as instead of a priority queue? (it can be iterated more easily)?